File: tools/bass-cover.py
"""Download YouTube audio and separate it into bass and no-bass stems."""

import argparse
import os
import re
import shlex
import subprocess
import sys
import time
from pathlib import Path

REPO_DIR = Path(__file__).resolve().parent.parent
MODELS = REPO_DIR / "data/onnx-lean"


def run(label, command):
    command_text = " ".join(shlex.quote(str(part)) for part in command)
    print(f"\n{'=' * 72}")
    print(f"STEP: {label}")
    print(f"{'=' * 72}")
    print(f"$ {command_text}", flush=True)

    start = time.perf_counter()
    subprocess.run(command, check=True)
    elapsed = time.perf_counter() - start

    print(f"{'-' * 72}")
    print(f"DONE: {label} ({elapsed:.2f}s)")
    print(f"{'-' * 72}", flush=True)


def time_slug(value):
    return f"{value:g}".replace(".", "p")


def slugify(value):
    value = value.lower()
    value = re.sub(r"[^a-z0-9]+", "-", value)
    return value.strip("-") or "song"


def youtube_url(value):
    if value.startswith(("http://", "https://")):
        return value
    return f"https://www.youtube.com/watch?v={value}"


def main():
    parser = argparse.ArgumentParser(
        description="Download YouTube audio, optionally trim it, and run Demucs bass separation."
    )
    parser.add_argument("youtube", help="YouTube ID or URL")
    parser.add_argument("--name", help="output basename")
    parser.add_argument("--start", type=float, help="trim start time in seconds")
    parser.add_argument("--end", type=float, help="trim end time in seconds")
    args = parser.parse_args()

    if args.end is not None:
        start = args.start or 0
        if args.end <= start:
            parser.error("--end must be greater than --start")

    if not MODELS.is_dir():
        sys.exit(f"missing {MODELS}\nbuild it first: pnpm build-model htdemucs_ft_bass")

    os.chdir(REPO_DIR)
    name = slugify(args.name or args.youtube)
    url = youtube_url(args.youtube)

    input_dir = Path("data/input")
    input_dir.mkdir(parents=True, exist_ok=True)

    source = input_dir / f"{name}.wav"
    run(
        "download audio",
        [
            "yt-dlp",
            "--no-playlist",
            url,
            "-x",
            "--audio-format",
            "wav",
            "-o",
            str(input_dir / f"{name}.%(ext)s"),
        ],
    )

    demucs_input = source
    if args.start is not None or args.end is not None:
        trim_parts = ["trim"]
        if args.start is not None:
            trim_parts.append(f"s{time_slug(args.start)}")
        if args.end is not None:
            trim_parts.append(f"e{time_slug(args.end)}")
        clip = input_dir / f"{name}-{'-'.join(trim_parts)}.wav"
        command = ["ffmpeg", "-y"]
        if args.start is not None:
            command.extend(["-ss", str(args.start)])
        if args.end is not None:
            command.extend(["-to", str(args.end)])
        command.extend(["-i", str(source), str(clip)])
        run("trim clip", command)
        demucs_input = clip

    output_dir = Path("data/output") / demucs_input.stem
    run(
        "separate bass stem",
        [
            "pnpm",
            "cli-separate",
            "--name",
            "htdemucs_ft",
            "--two-stems",
            "bass",
            "--method",
            "minus",
            demucs_input,
            output_dir,
        ],
    )

    print(f"\nOutput: {output_dir}/")


if __name__ == "__main__":
    main()

File: examples/bass-cover.ts
#!/usr/bin/env node

import { spawnSync } from "node:child_process";
import { existsSync, mkdirSync } from "node:fs";
import { basename, dirname, join, resolve } from "node:path";
import { performance } from "node:perf_hooks";
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";

const repoDir = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const models = join(repoDir, "data/onnx-lean");

function usage(): string {
  return `Usage: pnpm bass-cover <youtube-id-or-url> [options]

Download YouTube audio, optionally trim it, and separate bass.wav and no_bass.wav.

Options:
  --name <name>     Output basename
  --start <seconds> Trim start time
  --end <seconds>   Trim end time
  -h, --help        Show this help`;
}

function fail(message: string): never {
  console.error(message);
  process.exit(1);
}

function quote(value: string): string {
  return /^[a-zA-Z0-9_./:=+-]+$/.test(value)
    ? value
    : `'${value.replaceAll("'", `'\\''`)}'`;
}

function run(label: string, command: string, args: string[]): void {
  console.log(`\n${"=".repeat(72)}`);
  console.log(`STEP: ${label}`);
  console.log("=".repeat(72));
  console.log(`$ ${[command, ...args].map(quote).join(" ")}`);

  const start = performance.now();
  const result = spawnSync(command, args, { cwd: repoDir, stdio: "inherit" });
  if (result.error) {
    fail(`${label} failed: ${result.error.message}`);
  }
  if (result.status !== 0) {
    process.exit(result.status ?? 1);
  }

  console.log("-".repeat(72));
  console.log(
    `DONE: ${label} (${((performance.now() - start) / 1000).toFixed(2)}s)`,
  );
  console.log("-".repeat(72));
}

function numberOption(
  name: string,
  value: string | undefined,
): number | undefined {
  if (value === undefined) {
    return undefined;
  }
  const number = Number(value);
  if (!Number.isFinite(number)) {
    fail(`--${name} must be a number`);
  }
  return number;
}

const { values, positionals } = parseArgs({
  allowPositionals: true,
  options: {
    name: { type: "string" },
    start: { type: "string" },
    end: { type: "string" },
    help: { type: "boolean", short: "h" },
  },
});

if (values.help) {
  console.log(usage());
  process.exit(0);
}
if (positionals.length !== 1) {
  fail(usage());
}

const [youtube] = positionals;
const start = numberOption("start", values.start);
const end = numberOption("end", values.end);
if (end !== undefined && end <= (start ?? 0)) {
  fail("--end must be greater than --start");
}
if (!existsSync(models)) {
  fail(`missing ${models}\nbuild it first: pnpm build:model htdemucs_ft_bass`);
}

const slugify = (value: string): string =>
  value
    .toLowerCase()
    .replaceAll(/[^a-z0-9]+/g, "-")
    .replaceAll(/^-|-$/g, "") || "song";
const timeSlug = (value: number): string => String(value).replace(".", "p");
const name = slugify(values.name ?? youtube);
const url =
  youtube.startsWith("http://") || youtube.startsWith("https://")
    ? youtube
    : `https://www.youtube.com/watch?v=${youtube}`;
const inputDir = join(repoDir, "data/input");
mkdirSync(inputDir, { recursive: true });

const source = join(inputDir, `${name}.wav`);
run("download audio", "yt-dlp", [
  "--no-playlist",
  url,
  "-x",
  "--audio-format",
  "wav",
  "-o",
  join(inputDir, `${name}.%(ext)s`),
]);

let demucsInput = source;
if (start !== undefined || end !== undefined) {
  const trimParts = ["trim"];
  if (start !== undefined) {
    trimParts.push(`s${timeSlug(start)}`);
  }
  if (end !== undefined) {
    trimParts.push(`e${timeSlug(end)}`);
  }
  demucsInput = join(inputDir, `${name}-${trimParts.join("-")}.wav`);

  const ffmpegArgs = ["-y"];
  if (start !== undefined) {
    ffmpegArgs.push("-ss", String(start));
  }
  if (end !== undefined) {
    ffmpegArgs.push("-to", String(end));
  }
  ffmpegArgs.push("-i", source, demucsInput);
  run("trim clip", "ffmpeg", ffmpegArgs);
}

const outputDir = join(repoDir, "data/output", basename(demucsInput, ".wav"));
run("separate bass stem", "pnpm", [
  "cli-separate",
  "--name",
  "htdemucs_ft",
  "--two-stems",
  "bass",
  "--method",
  "minus",
  demucsInput,
  outputDir,
]);

console.log(`\nOutput: ${outputDir}/`);
